content/blog/kubernetes-anti-patterns/index.md

```diff
@@ -0,0 +1,256 @@
+---
+title: "Avoiding Kubernetes Anti-Patterns"
+date: 2020-10-01
+meta_desc: "Prevent Kubernetes anti-patterns with infrastructure as code."
+meta_image: anti-patterns.png
+authors:
+    - sophia-parafina
+tags:
+    - kubernetes
+    - anti-patterns
+---
+
+In software development, an anti-pattern is defined as an apparent solution that has unintended or negative consequences. The other side of anti-patterns is that they also offer solutions. Let's look at [container](https://codefresh.io/containers/docker-anti-patterns/) and [Kubernetes](https://medium.com/better-programming/10-antipatterns-for-kubernetes-deployments-e97ce1199f2d) anti-patterns and how to avoid them with infrastructure as code.
+
+<!--more-->
+
+## Hardcoding configuration in containers
+
+Containers are immutable, and a single image can be used throughout the software lifecycle, i.e., from dev, to test, and to production. However, developers frequently use an image built using artifacts and configurations unique to a stage in the software lifecycle. The result is that the image in dev or test is not the same as the image deployed in production.
+
+Best practice uses ConfigMaps, a Kubernetes native resource that doesn't require integration.   Because configuration is decoupled from the application, building the image at each stage is not necessary. Furthermore, ConfigMaps let you update when the application is running because the configuration is available during runtime.
+
+Let's look at an example where NginX is configured as a reverse-proxy that sends traffic to pulumi.github.io. The configuration is contained in a file, default.conf. Changing the file registers as a change in the ConfigMap's data triggering a rollout of the NginX Deployment. If we change .node.server and .serve.location.proxy_set_header to `google.com`, NginX will redirect requests to google.com. When we run `pulumi up`m the preview shows it will explicitly trigger a rollout in the Deployment by replacing the ConfigMap. In contrast, kubectl will silently sync the new version of the file to the containers.
```

Comment:
```suggestion
Let's look at an example where NginX is configured as a reverse-proxy that sends traffic to `pulumi.github.io`. The configuration is contained in a file, `default.conf`. Changing the file registers as a change in the ConfigMap's data triggering a rollout of the NginX Deployment. If we change `.node.server` and `.serve.location.proxy_set_header` to `google.com`, NginX will redirect requests to google.com. When we run `pulumi up`, the preview shows it will explicitly trigger a rollout in the Deployment by replacing the ConfigMap. In contrast, kubectl will update the ConfigMap, but changes are not automatically applied to running containers.
```